### PR TITLE
UART16550: Add zynq7000 implementation

### PR DESCRIPTION
**Pull request by agkaminski**
## Motivation and Context

PP-86

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zynq7000-zedboard).

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/uart16550/uarthw-zynq7000.c [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+	*p = val;
+
+	return;
+}
+
+
+char *uarthw_dump(void *hwctx, char *s, size_t sz)
+{
+	snprintf(s, sz, "base=0x%p irq=%d", (volatile void *)(((uarthw_ctx_t *)hwctx)->base), ((uarthw_ctx_t *)hwctx)->irq);
```
Comment (gerard5) [inline tty/uart16550/uarthw-zynq7000.c:82]: `irq` is unsigned, so maybe better is `%u` (for ia32 and riscv too).

File: tty/uart16550/uarthw-zynq7000.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+
+void uarthw_write(void *hwctx, unsigned int reg, uint8_t val)
+{
+	volatile uint32_t *p = (((uarthw_ctx_t *)hwctx)->base + reg);
+
+	*p = val;
+
+	return;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply tty/uart16550/uarthw-zynq7000.c]: Done

File: tty/uart16550/uarthw-zynq7000.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	snprintf(s, sz, "base=0x%p irq=%d", (volatile void *)(((uarthw_ctx_t *)hwctx)->base), ((uarthw_ctx_t *)hwctx)->irq);
+	return s;
+}
+
+
+unsigned int uarthw_irq(void *hwctx)
+{
+	uarthw_ctx_t *ctx = (uarthw_ctx_t *)hwctx;
+
+	return ctx->irq;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply tty/uart16550/uarthw-zynq7000.c]: Done

File: tty/uart16550/uarthw-zynq7000.c (hunk 1 above)
Comment (agkaminski) [reply tty/uart16550/uarthw-zynq7000.c]: Done

**Review by gerard5 — APPROVED** [review]
Looks good to me, btw. I verified on a real HW (i586 with an ISA card compatible with uart16550).

File: tty/uart16550/uart16550.c [hunk 4]
```diff
@@ -242,14 +242,16 @@ static void poolthr(void *arg)
 
 static int _uart_init(uart_t *uart, unsigned int uartn, unsigned int speed)
 {
-	unsigned int divisor = 115200 / speed;
+	unsigned int divisor;
 	libtty_callbacks_t callbacks;
 
-	int err = uarthw_init(uartn, uart->hwctx, sizeof(uart->hwctx));
+	int err = uarthw_init(uartn, uart->hwctx, sizeof(uart->hwctx), &divisor);
```
Comment (HBuczynski) [inline tty/uart16550/uart16550.c:248]: It seems that variable divisor has two faces. Value from `uarthw_init` is a clock, not a divisor ;) It is just a little misleading.

File: tty/uart16550/uart16550.c [hunk 5]
```diff
@@ -242,14 +242,16 @@ static void poolthr(void *arg)
...
 	if (err < 0) {
 		return err;
 	}
 
+	divisor /= 16 * speed;
```
Comment (HBuczynski) [inline tty/uart16550/uart16550.c:253]: Magic number `16`. I will appreciate a comment.

**Review by HBuczynski — COMMENTED** [review]
Each of the `uarthw-*.c` source file contains a duplication of struct `uarthw_ctx_t` declaration. What do You think about moving it to `uarhw.h` to be a part of the interface?

File: tty/uart16550/uart16550.c (hunk 4 above)
Comment (agkaminski) [reply tty/uart16550/uart16550.c:248]: I later modify this variable to get divisor. Yeah, maybe 2 variables would be most readable, but I don't find this very problematic. You're free to change this, if you wish

File: tty/uart16550/uart16550.c (hunk 5 above)
Comment (agkaminski) [reply tty/uart16550/uart16550.c:253]: Clock is divided internally by 16 (rx oversampling). Pls add comment if you find it necessary

**Comment by agkaminski** [comment]
uarthw_ctx_t is meant to be per arch to allow flexible implementation, I would opt to leave it as is

**Review by HBuczynski — APPROVED** [review]

**Pull request merged**